src/librustdoc/clean/mod.rs

```diff
@@ -1208,26 +1209,23 @@ fn resolve(cx: &DocContext, path_str: &str, is_val: bool) -> Result<(Def, Option
                     };
                     Ok((ty.def, Some(format!("{}.{}", out, item_name))))
                 } else {
-                    let is_enum = match ty.def {
-                        Def::Enum(_) => true,
-                        _ => false,
-                    };
-                    let elem = if is_enum {
-                        cx.tcx.adt_def(did).all_fields().find(|item| item.ident.name == item_name)
-                    } else {
-                        cx.tcx.adt_def(did)
-                              .non_enum_variant()
-                              .fields
-                              .iter()
-                              .find(|item| item.ident.name == item_name)
-                    };
-                    if let Some(item) = elem {
-                        Ok((ty.def,
-                            Some(format!("{}.{}",
-                                         if is_enum { "variant" } else { "structfield" },
-                                         item.ident))))
-                    } else {
-                        Err(())
+                    match cx.tcx.type_of(did).sty {
+                        ty::TyAdt(def, _) => {
+                            if let Some(item) = def.all_fields()
+                                                   .find(|item| item.ident.name == item_name) {
+                                Ok((ty.def,
+                                    Some(format!("{}.{}",
+                                                 if def.is_enum() {
+                                                     "variant"
```

Comment:
Don't think this is correct, it won't find variants... unless you want fields? This is confusing. Although, not any more wrong than before I guess.

Reply:
So we need the differentiation between `.non_enum_variant()` and `.all_fields()`?

Reply:
The issue is that in this code (and before), you're iterating over the fields (of a struct, enum, etc.), but if it's an enum, you're calling the result `"variant"` (when it's more accurately an `"enumfield"` or something). The question is whether you actually want fields or variants for enums. If you want fields, you should rename the string, but if you actually _are_ looking for variants, you should be iterating over `variants` if it's an enum, rather than `all_fields()`.

Reply:
I'm not sure what exactly you're looking for here, tbh.

Reply:
If the type has a field or a variant with a given name.